docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,210 @@
+There are many CI/CD tools out there: `GitHub Actions <https://github.com/features/actions>`_,
+`GitLab CI/CD <https://docs.gitlab.com/ee/ci/>`_, `Jenkins <https://jenkins.io/doc/>`_,
+`CircleCI <https://circleci.com/docs/2.0/>`_, etc. It is recommended to select one that 
+works well with your remote Git repository.
+
+
+Continuous Integration (CI)
+---------------------------
+
+Assistants are best improved with frequent, `incremental updates
+<https://rasa.com/docs/rasa-x/user-guide/improve-assistant/>`_.
+No matter how small a change is, you want to be sure that it doesn't introduce
+new problems or negatively impact the performance of your assistant. Most tests are
+quick enough to run on every change. However, you can set some more
+resource-intentsive tests to run only when certain files have been changed or
+when a certain tag is present.
+
+It is usually best to run CI checks on merge/pull request or on commit.
+
+.. contents::
+   :local:
+
+Validate Data and Stories
+#########################
+
+:ref:`Data validation <validate-files>` verifies that there are no mistakes or
+major inconsistencies in your domain file, NLU data, or story data. 
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings --max-history 5
+
+If data validation results in errors, training a model will also fail. By
+including the ``--fail-on-warnings`` flag, validation will also fail on
+warnings about problems that won't prevent training a model, but might indicate
+messy data, such as actions listed in the domain that aren't used in any
+stories. 
+
+Data validation includes :ref:`story structure validation <test-story-files-for-conflicts>`.
+Story validation checks if you have any
+stories where different bot actions follow from the same dialogue history.
+Conflicts between stories will prevent a model from learning the correct
+pattern for a dialogue. ``--max-history`` should be set to the value of ``max_history`` in your
+config for your memoization policy (which defaults to ``5``).
+
+Train a Model
+#############
+
+.. code-block:: bash
+
+   rasa train
+
+Training a model verifies that your NLU pipeline and policy configurations are
+valid and trainable, and it provides a model to use for test conversations. 
+If your the passes the CI tests, then you can also upload the trained model 
+to your server as :ref:`part of the continuous deployment
+process <uploading-a-model>`. 
+
+Test the Assistant
+##################
+
+Testing your trained model on :ref:`test conversations
+<end-to-end-testing>` is the best way to have confidence in how your assistant
+will act in certain situations. These stories, written in a modified story
+format, allow you to provide entire conversations and test that, given this
+user input, your model will behave in the expected manner. This is especially
+important as you start introducing more complicated stories from user
+conversations. End-to-end testing is only as thorough and accurate as the test
```

Comment:
Maybe split into two paragraphs? Potentially at `End-to-end testing is only...`. It is pretty big right now